bpf/lib/encrypt.h

```diff
@@ -291,18 +294,48 @@ ipsec_maybe_redirect_to_encrypt(struct __ctx_buff *ctx, __be16 proto,
 			 * large with no additional options for Geneve.
 			 */
 			l4_off = ETH_HLEN + ipv4_hdrlen(ip4);
-			if (!vxlan_get_inner_ipv4(data, data_end, l4_off,
-						  &ip4_inner))
-					return CTX_ACT_OK;
+			l2_inner_off = l4_off + sizeof(struct udphdr) +
+				       sizeof(struct vxlanhdr);
+
+			if (data + l2_inner_off + sizeof(struct ethhdr) >
+			    data_end)
+				return CTX_ACT_OK;
 
-			/* if this is already encrypted, just pass it to stack */
-			if (ip4_inner->protocol == IPPROTO_ESP)
+			eth_inner = (struct ethhdr *)(data + l2_inner_off);
+			switch (eth_inner->h_proto) {
+			case ETH_P_IP:
```

Comment:
```suggestion
			switch (eth_inner->h_proto) {
			case bpf_htons(ETH_P_IP):
```

Should we maybe flip this PR back to draft, until you've had opportunity to add sufficient testing?

Reply:
What extent of testing are you thinking? 

I know you have the idea for an integration test here, but we didnt do this with the original backport PR.

Reply:
I believe a bunch of BPF tests with the packets you plan to process in this function are needed. We already have similar tests for bpf_lxc, this would simply establish equivalent coverage.

We didn't add such tests initially, but we should have. And they would have easily caught this bug, instead of relying on reviewers to spot them.

Reply:
@julianwiedmann we also have a unit test just for `ipsec_maybe_redirect_to_encrypt` which can be updated to catch this same test. 

I'm not sure if this PR requires a full integration test of `to_netdev` or an update that throws vxlan traffic into this function.

Reply:
Updated with tunnel traffic tests of `ipsec_maybe_redirect_to_encrypt`.